lib/data/hash/map.ml

```diff
@@ -0,0 +1,84 @@
+open import "../../prelude.ml"
+open import "../../lua/conversion.ml"
+open import "../hash.ml"
+include import "../index.ml"
+
+private module Math = import "../../lua/math.ml"
+private module Array = import "../array.ml"
+
+private type pairs 'k 'v =
+  | Nil
+  | Cons of { k : 'k, v : 'v, next : pairs 'k 'v }
+
+private type container 'k 'v = Container of {
+  array : Array.array (pairs 'k 'v),
+  size : int,
+  initial : int,
+}
+
+type t 'k 'v = private HashMap of ref (container 'k 'v)
```

Comment:
This type is missing instances for `functor`, `foldable`, and `traversable`.

Reply:
Good point. I originally left it off, as functors on mutable types feels a little  weird, but we also do in for arrays after all.

Reply:
foldable` at least makes a lot of sense, but I think a copying `map` is worth having, at least.